src/software/ai/hl/stp/skill/pass/pass_skill_fsm.cpp

```diff
@@ -0,0 +1,187 @@
+    }
+
+    return ball.velocity().length() <
+           ai_config.ai_parameter_config().ball_is_kicked_m_per_s_threshold();
+}
+
+void PassSkillFSM::findPass(
+    const Update& event, boost::sml::back::process<DribbleSkillFSM::Update> processEvent)
+{
+    // Check if best_pass_so_far_ has been set BEFORE setting best_pass_so_far_,
+    // and only set SkillState IF best_pass_so_far_ has already been set.
+    //
+    // This ensures that a PassSkillFSM that has just been "reset" (i.e. constructed anew)
+    // will not immediately set the SkillState, and consequently override any SkillState
+    // changes made by another existing PassSkillFSM.
+
+    if (best_pass_so_far_)
+    {
+        event.common.set_skill_state(
+            {.pass_committed = false, .pass = best_pass_so_far_->pass});
+    }
+
+    best_pass_so_far_ = event.common.strategy->getBestPass();
+
+    // Update minimum pass score threshold. Wait for a good pass by starting out only
+    // looking for "perfect" passes (with a high score close to 1) and decreasing this
+    // threshold over time
+
+    const auto& ai_config           = event.common.strategy->getAiConfig();
+    const double abs_min_pass_score = ai_config.passing_config().abs_min_pass_score();
+    const double min_perfect_pass_score =
+        ai_config.passing_config().min_perfect_pass_score();
+    const double pass_score_ramp_down_duration =
+        ai_config.passing_config().pass_score_ramp_down_duration();
+
+    if (!pass_optimization_start_time)
+    {
+        pass_optimization_start_time = event.common.world_ptr->getMostRecentTimestamp();
+    }
+
+    time_since_commit_stage_start =
+        event.common.world_ptr->getMostRecentTimestamp() - *pass_optimization_start_time;
+    min_pass_score_threshold_ =
+        min_perfect_pass_score - std::min(time_since_commit_stage_start.toSeconds() /
+                                              pass_score_ramp_down_duration,
+                                          min_perfect_pass_score - abs_min_pass_score);
+
+    Point dribble_destination = event.common.world_ptr->ball().position();
+
+    // If the best pass so far has a less than "perfect" score,
+    // try dribbling to an open area
+    if (best_pass_so_far_->rating < min_perfect_pass_score)
+    {
+        dribble_destination = findKeepAwayTargetPoint(
+            *event.common.world_ptr, best_pass_so_far_->pass,
+            event.common.strategy->getAiConfig().passing_config());
+    }
+
+    Point receiver_point      = best_pass_so_far_->pass.receiverPoint();
+    Angle dribble_orientation = (receiver_point - dribble_destination).orientation();
+
+    DribbleSkillFSM::ControlParams control_params = {
+        .dribble_destination       = dribble_destination,
+        .final_dribble_orientation = dribble_orientation,
+        .excessive_dribbling_mode  = TbotsProto::ExcessiveDribblingMode::TERMINATE};
+
+    processEvent(DribbleSkillFSM::Update(control_params, event.common));
```

Comment:
lol after robocup we need to figure out how to cleanly separate out skills and tactics